test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/FilterAndOrderByFunctionalTests.cs

```diff
@@ -844,27 +844,31 @@ public void IsOfFunctionWithTwoParameters_WithoutSingleQuotesOnTypeParameter_Sho
         }
 
         [Theory]
-        [InlineData("isof(Fully.Qualified.Namespace.Pet1)", "Fully.Qualified.Namespace.Pet1")]
-        [InlineData("cast(Fully.Qualified.Namespace.HomeAddress)/City eq 'City1'", "Fully.Qualified.Namespace.HomeAddress")]
-        public void IsOfAndCastFunctionsWithSingleParameterWithoutSingleQuotes_WithIncorrectType_ThrowException(string filterQuery, string fullyQualifiedTypeName)
+        [InlineData("isof(Fully.Qualified.Namespace.Pet1)")]
+        [InlineData("isof(MyAddress,Fully.Qualified.Namespace.Pet1)")]
+        [InlineData("isof(null,Fully.Qualified.Namespace.Person)")]
+        [InlineData("isof('',Fully.Qualified.Namespace.Person)")]
+        public void IsOfFunctionsWithUnquotedTypeParameter_WithIncorrectType_DoesNotThrowException(string filterQuery)
         {
             // Arrange & Act
-            Action test = () => ParseFilter(filterQuery, HardCodedTestModel.TestModel, HardCodedTestModel.GetPersonType(), HardCodedTestModel.GetPeopleSet());
+            var exception = Record.Exception(() => ParseFilter(filterQuery, HardCodedTestModel.TestModel, HardCodedTestModel.GetPersonType(), HardCodedTestModel.GetPeopleSet()));
 
             // Assert
-            test.Throws<ODataException>(Strings.MetadataBinder_HierarchyNotFollowed(fullyQualifiedTypeName, "Fully.Qualified.Namespace.Person"));
+            Assert.Null(exception);
         }
 
         [Theory]
-        [InlineData("isof(MyAddress,Fully.Qualified.Namespace.Pet1)", "Fully.Qualified.Namespace.Pet1")]
-        [InlineData("cast(MyAddress,Fully.Qualified.Namespace.Employee)/WorkID eq 345", "Fully.Qualified.Namespace.Employee")]
-        public void IsOfAndCastFunctionsWithTwoParameterWhereTypeParameterIsWithoutSingleQuotes_WithIncorrectType_ThrowException(string filterQuery, string fullyQualifiedTypeName)
+        [InlineData("cast(Fully.Qualified.Namespace.HomeAddress)/City eq 'City1'")]
+        [InlineData("cast(MyAddress,Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        [InlineData("cast(null,Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        [InlineData("cast('',Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        public void CastFunctionWithUnquotedTypeParameter_WithIncorrectType_DoesNotThrowException(string filterQuery)

```

Comment:
What is the expected behaviour of `cast()` when the types are incorrect? If it doesn't throw an exception, what will happen? How is this scenario expected to be handled?

Reply:
This is what is happening currently with `cast` with `quoted type parameter`
- ODL will bind `literal type parameter token` to `ConstantNode` using [BindLiteral(LiteralToken literalToken)](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Binders/LiteralBinder.cs#L21). The AspNetCore library will get the `cast` node and tries to create LINQ Expression using [Expression BindCastSingleValue(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L562). It knows how to handle `ConstantNode`

For `cast` with `unquoted type parameter`
- ODL will bind `DottedIDentifierToken` as `SingleResourceCastNode` using a method I have added with this PR. To support this in AspNetCore, I will update function [Expression BindCastSingleValue(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L562) to add a condition to cast QueryNode to SingleResourceCastNode like what was done with the PR: https://github.com/OData/AspNetCoreOData/pull/1313

Reply:
Thanks for this added context. Can you also shed light on the behaviour with respect to handling of types where a cast is not possible? Ideally, from the user's endpoint the behaviour should be the same regardless of whether they user quoted or non-quoted syntax, and regardless of our internal implementation details which you have highlighted.

So could you clarify how we handle both supported and invalid cast scenarios in both quoted and unquoted variants? In which cases are we throwing an exception? Is the exception thrown consistent? Is that the expected behaviour?

My expectation is that `isof` should not throw an exception when a cast is invalid, but rather evaluate to a boolean that will perform the correct filter. I think `cast` throwing an exception when a cast is invalid is reasonable. What is not clear to me is whether such an exception should be thrown in ODL or AspNetCore. For the case of `cast` I would be cautious about changing the existing behaviour to avoid introducing a breaking change.

Reply:
With the current behavior, ODL does not throw exception or to be more specific, it doesn't check if target type is related to source type. It only bind the `quoted type parameters` to `constant node` and return the entire token as FunctionCall token. The `AspNetCore`, on the other end, will get the token and try to create LINQ expression. If expression is not possible, AspNetCore returns null [BindSingleResourceCastFunctionCall](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L655) will return `null` expression. 

Let's look at this query:
```
/products?$filter=cast('Microsoft.AspNetCore.OData.E2E.Tests.Cast.Category')/Name eq 'Cat
```

AspNetCore will try to relate `Entity Product` to `Entity Category` in [BindSingleResourceCastFunctionCall](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L655) and returns null expression if one is not assignable from the other.

And then [CreatePropertyAccessExpression](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L1114) will throw an expression when trying to get the Property Name from a null expression.

![image](https://github.com/user-attachments/assets/ab57583f-99fc-43bc-9a6c-502cf206e2ae)

An exception like below will be thrown:
```
 Message: 
System.ArgumentException : Instance property 'Name' is not defined for type 'System.Object' (Parameter 'propertyName')

  Stack Trace: 
Expression.Property(Expression expression, String propertyName)
QueryBinder.CreatePropertyAccessExpression(Expression source, QueryBinderContext context, IEdmProperty property, String propertyPath) line 1124
QueryBinder.BindPropertyAccessQueryNode(SingleValuePropertyAccessNode propertyAccessNode, QueryBinderContext context) line 360
QueryBinder.BindSingleValueNode(SingleValueNode node, QueryBinderContext context) line 147
QueryBinder.Bind(QueryNode node, QueryBinderContext context) line 73
QueryBinder.BindBinaryOperatorNode(BinaryOperatorNode binaryOperatorNode, QueryBinderContext context) line 453
QueryBinder.BindSingleValueNode(SingleValueNode node, QueryBinderContext context) line 132
QueryBinder.Bind(QueryNode node, QueryBinderContext context) line 73
FilterBinder.BindFilter(FilterClause filterClause, QueryBinderContext context) line 39
```